### ISSUE TITLE
ros-indigo-desktop-full

### ISSUE THREAD

**Issue by polygamma** [actionless/pikaur#53]
Hey!

Trying to install ros-indigo-desktop-full (https://aur.archlinux.org/packages/ros-indigo-desktop-full) yields that python2-catkin-pkg cannot be found in AUR and hence ros-indigo-desktop-full cannot be installed.

Two things: 

- It would be nice if pikaur would show that the missing dep is a dep of ros-indigo-catkin
- It actually is possible to know for pikaur that python2-catkin-pkg is being provided by python2-catkin_pkg since it's being listed explicitly in other packages of the dependency chain, hence it should be possible to solve the dependency chain for ros-indigo-desktop-full all in all

**Comment by actionless** [comment]
the thing is what search in AUR's rpc is not indexing Provides field so the only thing which i can do here -- get the full package list from AUR (like in `-Ss` without args) and next search through local 'mirror' of AUR db for such packages

**Comment by AladW** [comment]
Do any of you both know PHP? Sounds easier to file a patch to `aurweb` than come up with a good workaround here :smile:

**Comment by actionless** [comment]
i don't know php but i think i can add to RPC one more filter criteria, as `name` and `name-desc` as it have now to have also either `provides` or `name-desc-provides` (probably someone else should come up with a better name )

**Comment by actionless** [comment]
i didn't noticed your edits on the original comment:

regarding local mirror, yes, see here: https://github.com/actionless/pikaur/blob/master/pikaur/aur.py#L306-L310

**Comment by polygamma** [comment]
@actionless 
Reference: https://github.com/AladW/aurutils/issues/10#issuecomment-370270445

The first part explains how aurman fetches packages and thus gets the needed information. May be interesting for you.

**Comment by actionless** [comment]
@AladW i had a moment to have a look on https://git.archlinux.org/aurweb.git/tree/web/lib/aurjson.class.php it doesn't look complicated, what is the procedure to submit a patch?

**Comment by actionless** [comment]
or alternative approach to fixing that on AUR RPC side could be when searching by name match not only Name field but also Provides field

**Comment by AladW** [comment]
@actionless the procedure is to send your patch to [aur-dev](https://lists.archlinux.org/listinfo/aur-dev) where it can be merged after discussion. If you haven't used `git send-email` before, see https://www.freedesktop.org/wiki/Software/PulseAudio/HowToUseGitSendEmail/

**Comment by actionless** [comment]
so i can't start a discussion topic without subscribing to the whole mail list?

**Comment by AladW** [comment]
I don't recall if you have to be actually subscribed to send messages. If you do and you don't want to recieve other topics, you can disable recieving them in the mailman settings.

**Comment by actionless** [comment]
ok, i'll try doing that tonight or so

**Comment by actionless** [comment]
@AladW how do you think which of the approaches will be more acceptable:

1) adding `name-desc-provides` filter
2) adding `provides` filter
3) when searching by `name` match not only Name field but also Provides field
4) mb smth else?

**Comment by AladW** [comment]
Personally I would prefer the third option, since I'm not sure what the benefits of an additional filter over a transparent approach are. I don't speak for the AUR development team though, so I'd propose these options on aur-dev as well. (the list is equally suited for discussion and posting of patches)

**Comment by actionless** [comment]
@AladW i've submitted a message there but it didn't appeared here:
https://lists.archlinux.org/pipermail/aur-dev/2018-March/date.html

could it be because my e-mail address is not subscribed to the mail list?

**Comment by AladW** [comment]
Possible, you should have had a rejection mail in that case. Try subscribing then posting it (as mentioned, you can opt-out to receive messages from the list while still being subscribed)

**Comment by actionless** [comment]
i haven't received rejection neither, but i'll try to subscribe and send the message again

**Comment by AladW** [comment]
...please don't.

**Comment by actionless** [comment]
@AladW see last message from Florian in the mail list -- such kind of control freaks is the reason why i didn't wanted to touch the mail list even with a long stick

**Comment by AladW** [comment]
You shouldnt take it personally - the Arch mailing lists have a long history with email issues so people will remind you immediately. Anyway as you have may have noticed there's a positive response to the proposed changes.

**Comment by actionless** [comment]
don't worry, i'm not giving up ;-)

just mentioned why i so wanted to avoid that initially

**Comment by XavierCLL** [comment]
There are `provides` problems too when pikaur try to delete the make dependencies  after finished making a package:
![screenshot_20180308_144903](https://user-images.githubusercontent.com/11801453/37172836-f2995b18-22df-11e8-973f-e4cbe07ecc4e.png)
(here python-distribute is provider by python-setuptools)

**Comment by actionless** [comment]
@XavierCLL could you check this again with the latest git version of pikaur?

**Comment by XavierCLL** [comment]
not yet, same issue using Pikaur 0.8+19+g094ee87

**Comment by actionless** [comment]
@XavierCLL pushed more fixes

**Comment by XavierCLL** [comment]
perfect, the last commit fixed the problem, thanks @actionless

**Comment by polygamma** [comment]
@actionless, I've looked at the following patch: https://transfer.sh/L3H4Z/0001-feat-rpc-return-providers-packages-when-querying-by-.patch

I think it's actually a very bad idea to alter the existing `name` and `name-desc` fields. People are already working with that, and if you change it, many things might break, since you are changing what those fields do. Everybody, who is working with AurJson needs to change existing code, to adapt to the new usage of the fields. Such API changes should really never, ever happen.

**Comment by actionless** [comment]
i did proposed another alternatives but this was chosen, feel free to file a patch implementing that in other way

**Comment by AladW** [comment]
Simple solution: bump the RPC version to 6. People who want the old API can keep using 5.

**Comment by Morganamilo** [comment]
What about adding this to `info`? `info` allows multiple arguments, so say a package you're fetching the dependencies for has 10 AUR dependencies. Before you would just do 1 info request with those 10  packages. Now you would have to do an individual search request for every dependency then get info from then. Which brings the request count up to 11.

Not to mention search is for, well searching. It does not look for exact names. So when you have the dependency `foo` and you want to find packages that also provide `foo` after you preform the search on the rpc you would have to check each package to see if they actually provide `foo`, or just happen to have a similar name.

I think this does belong in `name` and `name-desc` for `-Ss` type searching but for actual dependency resolving this should also also be added to `info`.

**Comment by actionless** [comment]
time passed from submitting patch for AUR: 10 days
time necessary to implement in python parsing of such a simple html page: 10 minutes or so

i'm putting this to 0.10 milestone (magic number here is not intentional ;-) ) and so if it won't be any progress till then will merge html-based solution

**Comment by afg984** [comment]
Late to the discussion here, and I don't actually know how to post on the mailing list, so I'm posting my thoughts here.

I think we'd better introduce a new `type=provides` or `search_by=provides`. My understanding is that there's no way in the RPC we can get a list of packages that are capable of providing the dependency. Putting it into the search means AUR helpers will have to verify whether the search results actually provides the required dependency afterwards.

@actionless How did you get the testing server working? I'm interested in creating a patch but have trouble getting the database set up.

What I've done in the config is this but I think it's wrong someway as the web page shows `Error - Could not connect to AUR database`:

```
[database]
backend = sqlite
name = /home/afg/aurweb/aurweb.sqlite3
```

Parsing the html just seems soooo easy compared to this though :(

**Comment by actionless** [comment]
@afg984 i was also complaining on the maillist about having readme and not working example config files (which have to be edited manually according to the readme to get them work :D) instead of just small clean Docker/Vagrant file which instantly works as expected and serves as reference setup example. that's just so not right in 2018.

regarding your problem specifically -- mb you haven't enabled pdo module for sqlite

**Comment by Morganamilo** [comment]
To be more specific, uncomment `extention=pdo_sqlite` in `/etc/php/php.ini`.

**Comment by AladW** [comment]
>time necessary to implement in python parsing of such a simple html page: 10 minutes or so

Yes, and that also means that everyone else who wants this would also have to implement expensive and ugly HTML parsing. A patch means everyone can take advantage of the new and improved API.

PS. I agree with @Morganamilo on adding this to `info`

**Comment by actionless** [comment]
there's still no reviews after i've submitted it in the readable reviews, nuff said

**Comment by AladW** [comment]
Eh. You submitted it as a link to some 10 mb javascript page that doesn't even display the actual file. I linked the standard procedure over 2 weeks ago in https://github.com/actionless/pikaur/issues/53#issuecomment-370431173, not sure why it's so hard to follow.

**Comment by afg984** [comment]
Thanks for the hints, got local aurweb running.

I'm all for having the information in `info`.
Still hoping we can `search_by=provides` for somewhat reason we use a PKGBUILD that's not obtained from the AUR but want resolve dependencies properly. It's not possible without crawling the whole AUR.

For the html stuff, I think fetching the information from the web and also comparing the results with the RPC is quite reliable. A xpath query of `'//*[@id="pkgdepslist"]/li'` will just do the trick.

**Comment by actionless** [comment]
@AladW i've already explained what that procedure is not designed for accessing not from the personal pc

btw transfer.sh is opening files with noscript enabled

but anyway i already what if i will upload it differently those guys will find some new way to complain and write 9000 lines of text intead of writing 3 lines of code to just fix the AUR

**Comment by Morganamilo** [comment]
I'll submit it right now if you like. I would like this to get reviewed so that we can progress with this. When your patch is all fixed up and finalised I can look into making a version of it for info.

**Comment by actionless** [comment]
yup, count the patch as public domain

**Comment by Morganamilo** [comment]
Sent. Hope it goes somewhere so I can translate it to `info`.

**Comment by actionless** [comment]
@Morganamilo @AladW any move on this on AUR side? i am about to close the milestone soon

**Comment by AladW** [comment]
Upstream agrees it's a good thing to add. The only remaining point to argue about is what to do with `info` queries.

https://lists.archlinux.org/pipermail/aur-dev/2018-March/004495.html

**Comment by actionless** [comment]
i think if it will be one api endpoint for both search and info it wouldn't be such design decision problem as now

imho instead of having that branching in code to support every request argument explicitly it could be better if it will be just handling any of fields available for search and just abstract api endoint, which takes one or more fieldnames and their values and doing search. i think such implementation will be the same SLOC as it is now or so but support all the fields for querying instead of few fixed ones

**Comment by AladW** [comment]
Until someone implements this, I've thought of an alternate approach: just retrieve ALL metadata from the RPC and store it locally, perhaps in some modified form. As this is ~20mb of JSON, this should be done manually through `--updatedb` or some such flag. This data could then be used to find the correct providers (and much more, such as full regex support for searches).

**Comment by actionless** [comment]
why do you think it will be better than parsing a web page?

**Comment by actionless** [comment]
btw i've just realized what retrieving of all aur metadata is already implemented in `-Ss #without args`

what is missing is only saving/loading that metadata to disk

>--updatedb
i was more into doing that on `--refresh`, because it seems to be almost the same thing from the user point of view, and so it could be controlled with additional `--aur` or `--repo` options (as it's done now in, for example, `-Ss`)

**Comment by AladW** [comment]
Because parsing the webpage for 50 000 packages takes an absurd amount of time, while doing so over the RPC doesn't? Anyway it should be a separate flag since it puts severe tax on the AUR, especially if users do an -Sy every 10 seconds.

**Comment by Morganamilo** [comment]
I think the point is to just parse the web pages to be installed. For example `pikaur -S pikaur` would fetch the deps from:
![image](https://user-images.githubusercontent.com/16593899/38523846-a6a56552-3c44-11e8-93c7-6816d40f2e94.png)

No need to go through 50k pages.

**Comment by AladW** [comment]
Well yes, but I was thinking of a more general solution here than ad-hoc parsing of the web interface.

**Comment by Morganamilo** [comment]
I do agree it's not a good idea, just pointing out it is feasible to implement. Having a cache of all package info seems like a better idea but still not something I like the thought of.

**Comment by actionless** [comment]
i would like the idea of syncing whole AUR database without any extra discussion if it would be possible to download it in couple of seconds as a single .gz from the server (currently querying all the AUR packages from RPC takes for me about 30 seconds)

**Comment by afg984** [comment]
FYI, I'm going to work on the RPC side once I get rid of the exams. I estimate it would take two weeks before the patches are ready though.

**Comment by actionless** [comment]
@Morganamilo do you know if there is any movement on the patch?

**Comment by Morganamilo** [comment]
It's been a while so I've forgotten the specifics. I submitted the patch on your behalf, they discussed it a bit, agreed it's a good idea but wanted some stuff changed, cant really remember.

I'm no php dev though. And I don't think they want to write it either. So it's up to you really.

**Comment by actionless** [comment]
but overall i have quite a weird impression of guys who not lazy to write tons of text but lazy to fix couple of lines of their own php code.. :confused:

**Comment by Morganamilo** [comment]
I think that's unfair. If I made a half done pull request to Pikaur that you wanted changed, would your first response be to give feedback or just finish it yourself?

**Comment by actionless** [comment]
in combination with blank arch container it should give me the expected result, so i can put it somewhere in `./maintenance_scripts` dir of the repo not to loose in case if i'll give it a try (for the upcoming week or two i am planning to concentrate on `oomox` app, so will get back to this problem when will be preparing 0.12/1.0 release of `pikaur`)

or mb @afg984's exams will finish til then :-)

**Comment by afg984** [comment]
The exams were over already, I was just overwhelmed and completely forgot about this :(

**Comment by afg984** [comment]
Hi all, I think I've finished most of the RPC, which addresses the issue of not being possible to query a dependency from the RPC. The design mostly follows the [one](https://lists.archlinux.org/pipermail/aur-dev/2018-March/004498.html) discussed on the mailing list.

[Here's](https://github.com/afg984/aurweb/wiki/aurweb-RPC-Interface-(v6-draft)) the draft document of the v6 RPC and [here's](https://github.com/afg984/aurweb/tree/v6rpc) the code.
I'd like to ask you to review the document and see if there are any problems of the design or any wording in the text that can be make better.

I've set up a testing site at http://140.114.91.172:8000/ using dummy data. I haven't done throughout testing yet, so please report if you've found any bugs.

Things that are not done yet:
* commit [1ff40987](https://git.archlinux.org/aurweb.git/commit/?id=1ff409874eda74dfc8c875e9fc1d0d512127746e) implemented search by depends, checkdepends, optdepends. They're not supported yet. The version (4.7.0) affected by that change is not deployed though.
* This was possible in v5:
  > If maintainer search is performed and the search argument is left empty, a list of orphan packages is returned.

**Comment by Morganamilo** [comment]
The way I do dependency resolving is by aggregating all dependencies into one rpc query. Because info supports multiple arguments at once it can just dump all the packages I want to resolve in one rpc request which is both fast and less intensive on the AUR.

It looks like you've only implemented provider finding on search. So say I have 10 dependencies I want to resolve `dep[0-9]`.

Previously to resolve them I would just do:

```
rpc.info(dep[0-9])
```

That's one request, all is good.

But with this I would have to do something along the lines of:

```
rpc.searchProvides(dep[0])
rpc.searchProvides(dep[0])
rpc.searchProvides(dep[0])

**Comment by afg984** [comment]
> It looks like you've only implemented provider finding on search. So say I have 10 dependencies I want to resolve dep[0-9].

This is definitely a bug that I overlooked.

**Comment by afg984** [comment]
@Morganamilo I just found I misunderstood php's query argument handling,
so the correct way to query multiple dependencies would be:
`arg[]=dep0&arg[]=dep1&arg[]=dep2`
(this is same for the by argument, for example: `by[]=name&by[]=provides`)

Example: http://140.114.91.172:8000/rpc/?v=6&type=info&by[]=name&by[]=provides&arg[]=wimps&arg[]=scoffers

**Comment by Morganamilo** [comment]
Ah right, provides is implemented in info as well as search. In that case I have no complaints, nice work.

**Comment by Morganamilo** [comment]
By the way when are you planning to submit this to aur-dev? Unless you already have and I missed it.

**Comment by afg984** [comment]
I haven't submitted it yet.

There are a few things that's holding me:
* I think the code changes / organization now is not easy to review
* The features that are ignored in v5
* Should we support multiple `arg[]` in `search` to perform a AND operation? (I think the SQL would look scary though)
  For example `pacman -Ss python audio` searches for something related to `python` and `audio`

Or should I just submit a patch and ask the official decision, before moving on? 

BTW, I've updated the document page to reflect the correct syntax.

**Comment by Morganamilo** [comment]
>I think the code changes / organization now is not easy to review

For sure, your code is all one commit with no descriptive tittle nor description.

>The features that are ignored in v5

What do you mean by this exactly?

>Should we support multiple arg[] in search to perform a AND operation?

This would be nice, but it's not really related to this. I would just focus on the provides code for now.

>Or should I just submit a patch and ask the official opinion on this, before moving on?

Nothing wrong with submitting a [WIP] patch for feedback. Saves you from writing code that ends up being unwanted.

**Comment by afg984** [comment]
>> I think the code changes / organization now is not easy to review

> For sure, your code is all one commit with no descriptive tittle nor description.

I think the bigger problem is, the old structure reuses the search() and multiinfo() constructs over RPC versions. v6 looks like a full rewrite and is particularly difficult to track down what has changed.

The reason I did that is due to search and info being the same concept in v6's design, with the only difference being substring (search) vs exact match (info).

>> The features that are ignored in v5

> What do you mean by this exactly?

Two working features mentioned in https://github.com/actionless/pikaur/issues/53#issuecomment-393601334 are no longer working in my v6.

> I would just focus on the provides code for now.

I think adding that later wouldn't break v6, so yes I'll focus on provides.

> Nothing wrong with submitting a [WIP] patch for feedback. Saves you from writing code that ends up being unwanted.

Sounds reasonable.

**Comment by afg984** [comment]
@Morganamilo 
I believe this was brought up before.
If package A depends on `java-environment>=8`
and 
package B provides `java-environment=9`,
how would you resolve this dependency?

`java-environment=9` seems to be stored in the database as is.

**Comment by Morganamilo** [comment]
The rpc ignores versioned deps to it might be fine to just totally return all providers of `java-environment`. Although you did mention there was a `{check,opt,make,}depends` support added to the rpc. Do these methods support versioned deps? Just do what they do.

**Comment by Morganamilo** [comment]
>What do you mean?

What I meant is, there is currently no way to search for versioned depstrings using the RPC. `rpc.Info("linux>4")` won't work.

>They do not support versioned deps.

Then I guess provides shouldn't support it either. You would have to do search for `java-environment` instead of `java-environment>8` which is consistent with how the rest of the system works.

**Comment by Morganamilo** [comment]
I should add these are just my opinions. These questions might be better suited for aur-dev.

**Comment by afg984** [comment]
> Then I guess provides shouldn't support it either. You would have to do search for java-environment instead of java-environment>8 which is consistent with how the rest of the system works.

Yes, what I meant was `type=info&by[]=provides&arg=[]java-environment` maybe should also return packages that defines `provides=(java-environment=9)`. My v6 only returns exact match.

> I should add these are just my opinions. These questions might be better suited for aur-dev.

Sure.

**Comment by Morganamilo** [comment]
>Yes, what I meant was type=info&by[]=provides&arg=[]java-environment maybe should also return packages that defines provides=(java-environment=9). My v6 only returns exact match.

Oh right yeah, for sure should return those packages too.

**Comment by audeoudh** [comment]
Hi,

Another example of non-working use-case : asking pikaur for installing `msp430-elf-gcc` will ask for its dependencies : among others `msp430-elf-newlib`, and then `msp430-elf-gcc-stage1`, in conflict with `msp430-elf-gcc` itself (so pikaur refuses to install anything)…  however, `msp430-elf-gcc` already provides `msp430-elf-gcc-stage1`.

Any news about this issue ?
